README.adoc
Crypt-DRBG
==========

== Overview

Crypt-DRBG is a set of fast, cryptographically-secure PRNGs.  It can be used
either as a supplement to the system PRNG (e.g. /dev/random or /dev/urandom) or
where reproducible results are sometimes needed (e.g. in unit tests).
It also provides bias-free selection of bytes or array elements.

The algorithms implemented are the NIST HMAC_DRBG and Hash_DRBG.  Unit tests are
included to ensure conformance with the specification.
The HMAC implementation (`Crypt::DRBG::HMAC`)
is the preferred implementation for speed reasons.
Both algorithms are generally believed to be secure.

== Dependencies

Crypt::DRBG requires only core Perl on Unix systems with a random device.

Digest::SHA is required, so if you're using an older version of Perl where that
isn't included in core, you'll need to install it.

Performance of the Hash DRBG and unit tests will be improved if
Math::BigInt::GMP is available.

Finally, Crypt::URandom will be used if it's available; this is required for
Windows machines when not using Cygwin.

== Installation

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

== License and Copyright

Copyright (C) 2015 brian m. carlson

This program is distributed under the MIT (X11) License:
http://www.opensource.org/licenses/mit-license.php

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.



Changes
Revision history for Crypt-DRBG

0.000003 2017-11-13
         * Make Crypt::DRBG::Hash constant time.
         * Mark GitHub as the source for issues.

0.000002 2016-02-03
         * Don't call new on Math::BigInt unnecessarily.
         * Fix test failure on old Digest::SHA.
         * Fix breakage on older versions of Math::BigInt.

0.000001 2015-11-28
         * Initial release.


t/lib/Test/DRBG.pm
package Test::DRBG;

use strict;
use warnings;

use FindBin;

use IO::File;
use Test::More;

sub run_tests {
	my ($name) = @_;
	my $fh = IO::File->new("$FindBin::Bin/support/${name}_DRBG.txt", 'r');

	my $algo;
	my $chunk;
	my @insns;
	my $set = 0;
	while (<$fh>) {
		if (/^\[SHA-([\d\/]+)\]/) {
			@insns = ();
			$algo = $1;
		}
		elsif (/^\[ReturnedBitsLen = (\d+)\]/) {
			$chunk = $1 / 8;
		}
		elsif (/^COUNT = (\d+)/) {
			next if $algo eq '1';
			run_test($name, \@insns);
			@insns = ();
			push @insns, {
				insn => 'setup',
				algo => $algo,
				count => $1,
				set => $set++,
			};
		}
		elsif (/^EntropyInput = ([0-9a-f]+)/) {
			push @insns, {insn => 'param', k => 'seed', v => pack("H*", $1)};
		}
		elsif (/^Nonce = ([0-9a-f]+)/) {
			push @insns, {insn => 'param', k => 'nonce', v => pack("H*", $1)};
		}
		elsif (/^PersonalizationString = ([0-9a-f]*)/) {
			push @insns, {
				insn => 'param',
				k => 'personalize',
				v => pack("H*", $1)
			},
			{
				insn => 'instantiate'
			};
		}
		elsif (/^\s+(Key|V|C)\s+=\s+([0-9a-f]+)/) {
			my $var = lc substr($1, 0, 1);
			push @insns, {insn => 'check', k => $var, v => pack("H*", $2)};
		}
		elsif (/^AdditionalInput = ([0-9a-f]*)/) {
			my $arg = pack('H*', $1 || '');
			push @insns, {insn => 'generate', n => $chunk, arg => $arg};
		}
		elsif (/^ReturnedBits = ([0-9a-f]+)/) {
			push @insns, {insn => 'data', arg => pack("H*", $1)};
		}
	}
	return;
}

sub is_hex {
	my ($got, $expected, $desc) = @_;

	$got = unpack('H*', $got);
	$expected = unpack('H*', $expected);

	return is($got, $expected, $desc);
}

sub run_test {
	my ($name, $insns) = @_;

	return unless @$insns;

	my $setup = shift @$insns;
	die unless $setup->{insn} eq 'setup';
	my $desc = "SHA-$setup->{algo}, count $setup->{count}, set $setup->{set}";

	subtest $desc => sub {
		my $drbg;
		my $data;
		my %params = (algo => $setup->{algo});
		while (my $insn = shift @$insns) {
			my $type = $insn->{insn};
			if ($type eq 'param') {
				$params{$insn->{k}} = $insn->{v};
			}
			elsif ($type eq 'instantiate') {
				$drbg = "Crypt::DRBG::$name"->new(%params);
				isa_ok($drbg, "Crypt::DRBG::$name");
				is($drbg->{algo}, $setup->{algo}, "Object has right algo");
			}
			elsif ($type eq 'check') {
				my $state = $drbg->{state};
				my $key = $insn->{k};
				is_hex($state->{$key}, $insn->{v}, "Value $key matches");
			}
			elsif ($type eq 'generate') {
				$data = $drbg->generate($insn->{n}, $insn->{arg} || undef);
				is(length($data), $insn->{n}, "Got $insn->{n} bytes");
			}
			elsif ($type eq 'data') {
				is_hex($data, $insn->{arg}, "Got expected data");
			}
			else {
				die "Bad insn $type";
			}
		}
	};

	return;
}

1;


lib/Crypt/DRBG/HMAC.pm
package Crypt::DRBG::HMAC;

use 5.006;
use strict;
use warnings;

use parent 'Crypt::DRBG';

use Digest::SHA ();

=head1 NAME

Crypt::DRBG::HMAC - Fast, cryptographically secure PRNG

=head1 SYNOPSIS

	use Crypt::DRBG::HMAC;

	my $drbg = Crypt::DRBG::HMAC->new(auto => 1);
	my $data = $drbg->generate(42);
	... # do something with your 42 bytes here

	my $drbg2 = Crypt::DRBG::HMAC->new(seed => "my very secret seed");
	my $data2 = $drbg->generate(42);

=head1 DESCRIPTION

Crypt::DRBG::HMAC is an implementation of the HMAC_DRBG from NIST SP800-90A.  It
is a fast, cryptographically secure PRNG.  By default, it uses HMAC-SHA-512.

However, if provided a seed, it will produce the same sequence of bytes I<if
called the same way each time>.  This makes it useful for simulations that
require good but repeatable random numbers.

Note, however, that due to the way the DRBGs are designed, making a single
request and making multiple requests for the same number of bytes will result in
different data.  For example, two 16-byte requests will not produce the same
values as one 32-byte request.

This class derives from Crypt::DRBG, which provides several utility functions.

=head1 SUBROUTINES/METHODS

=head2 Crypt::DRBG::HMAC->new(%params)

Creates a new Crypt::DRBG::HMAC.

%params can contain all valid values for Crypt::DRBG::initialize, plus the
following.

=over 4

=item algo

The algorithm to use for generating bytes.  The default is "512", for
HMAC-SHA-512.  This provides optimal performance for 64-bit machines.

If Perl (and hence Digest::SHA) was built with a compiler lacking 64-bit integer
support, use "256" here.  "256" may also provide better performance for 32-bit
machines.

=item func

If you would like to use a different hash function, you can specify a function
implemeting HMAC for your specific algorithm.  The function should take two
arguments, the value and the key, in that order.

For example, if you had C<Digest::BLAKE2> and C<Digest::HMAC> installed, you
could do the following to use BLAKE2b:

	my $func = sub {
		return Digest::HMAC::hmac(@_, \&Digest::BLAKEx::blake2b, 128);
	};
	my $drbg = Crypt::DRBG::HMAC->new(auto => 1, func => $func;
	my $data = $drbg->generate(42);

Note that the algo parameter is still required, explicitly or implicitly, in
order to know how large a seed to use.

=back

=cut

sub new {
	my ($class, %params) = @_;

	$class = ref($class) || $class;
	my $self = bless {}, $class;

	my $algo = $self->{algo} = $params{algo} || '512';
	$algo =~ tr{/}{}d;
	$self->{s_func} = ($params{func} || Digest::SHA->can("hmac_sha$algo")) or
		die "Unsupported algorithm '$algo'";
	$self->{seedlen} = $algo =~ /^(384|512)$/ ? 111 : 55;
	$self->{reseed_interval} = 4294967295; # (2^32)-1
	$self->{bytes_per_request} = 2 ** 16;
	$self->{outlen} = substr($algo, -3) / 8;
	$self->{security_strength} = $self->{outlen} / 2;
	$self->{min_length} = $self->{security_strength};
	$self->{max_length} = 4294967295; # (2^32)-1

	$self->initialize(%params);

	return $self;
}

sub _seed {
	my ($self, $seed) = @_;

	my $k = "\x00" x $self->{outlen};
	my $v = "\x01" x $self->{outlen};
	$self->{state} = {k => $k, v => $v};
	return $self->_reseed($seed);
}

sub _reseed {
	my ($self, $seed) = @_;

	$self->_update($seed);
	$self->{reseed_counter} = 1;
	return 1;
}

sub _update {
	my ($self, $seed) = @_;

	my $data = defined $seed ? $seed : '';

	my $func = $self->{s_func};
	my $state = $self->{state};
	my ($k, $v) = @{$state}{qw/k v/};
	$k = $func->("$v\x00$data", $k);
	$v = $func->($v, $k);
	if (defined $seed) {
		$k = $func->("$v\x01$data", $k);
		$v = $func->($v, $k);
	}
	@{$state}{qw/k v/} = ($k, $v);
	return 1;
}

=head2 $drbg->generate($bytes, $additional_data)

Generate and return $bytes bytes.  $bytes cannot exceed 2^16.

If $additional_data is specified, add this additional data to the DRBG.

=cut

sub _generate {
	my ($self, $len, $seed) = @_;

	$self->_check_reseed($len);
	$self->_update($seed) if defined $seed;

	my $count = ($len + ($self->{outlen} - 1)) / $self->{outlen};
	my $data = '';
	my $state = $self->{state};
	my ($k, $v) = @{$state}{qw/k v/};
	my $func = $self->{s_func};
	for (1..$count) {
		$v = $func->($v, $k);
		$data .= $v;
	}
	$self->{reseed_counter}++;
	@{$state}{qw/k v/} = ($k, $v);
	$self->_update($seed);
	return substr($data, 0, $len);
}

=head1 AUTHOR

brian m. carlson, C<< <sandals at crustytoothpaste.net> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-crypt-drbg at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Crypt-DRBG>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc Crypt::DRBG::HMAC


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-DRBG>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Crypt-DRBG>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Crypt-DRBG>

=item * Search CPAN

L<http://search.cpan.org/dist/Crypt-DRBG/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2015 brian m. carlson.

This program is distributed under the MIT (X11) License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.


=cut

1; # End of Crypt::DRBG::HMAC


lib/Crypt/DRBG/Hash.pm
package Crypt::DRBG::Hash;

use 5.006;
use strict;
use warnings;

use parent 'Crypt::DRBG';

use Digest::SHA ();

=head1 NAME

Crypt::DRBG::Hash - Fast, cryptographically secure PRNG

=head1 SYNOPSIS

	use Crypt::DRBG::Hash;

	my $drbg = Crypt::DRBG::Hash->new(auto => 1);
	my $data = $drbg->generate(42);
	... # do something with your 42 bytes here

	my $drbg2 = Crypt::DRBG::Hash->new(seed => "my very secret seed");
	my $data2 = $drbg->generate(42);

=head1 DESCRIPTION

Crypt::DRBG::Hash is an implementation of the Hash_DRBG from NIST SP800-90A.  It
is a fast, cryptographically secure PRNG.  By default, it uses SHA-512.

However, if provided a seed, it will produce the same sequence of bytes I<if
called the same way each time>.  This makes it useful for simulations that
require good but repeatable random numbers.

Note, however, that due to the way the DRBGs are designed, making a single
request and making multiple requests for the same number of bytes will result in
different data.  For example, two 16-byte requests will not produce the same
values as one 32-byte request.

This class derives from Crypt::DRBG, which provides several utility functions.

=head1 SUBROUTINES/METHODS

=head2 Crypt::DRBG::Hash->new(%params)

Creates a new Crypt::DRBG::Hash.

%params can contain all valid values for Crypt::DRBG::initialize, plus the
following.

=over 4

=item algo

The algorithm to use for generating bytes.  The default is "512", for
SHA-512.  This provides optimal performance for 64-bit machines.

If Perl (and hence Digest::SHA) was built with a compiler lacking 64-bit integer
support, use "256" here.  "256" may also provide better performance for 32-bit
machines.

=item func

If you would like to use a different hash function, you can specify a function
implemeting your specific algorithm.

For example, if you had C<Digest::BLAKE2> installed, you could do the following
to use BLAKE2b:

	my $drbg = Crypt::DRBG::Hash->new(
		auto => 1,
		func => \&Digest::BLAKE2::blake2b,
		algo => 512
	);
	my $data = $drbg->generate(42);

Note that the algo parameter is still required, in order to know how large a
seed to use.

=back

=back

=cut

sub new {
	my ($class, %params) = @_;

	$class = ref($class) || $class;
	my $self = bless {}, $class;

	my $algo = $self->{algo} = $params{algo} || '512';
	$algo =~ tr{/}{}d;
	$self->{s_func} = ($params{func} || Digest::SHA->can("sha$algo")) or
		die "Unsupported algorithm '$algo'";
	$self->{seedlen} = $algo =~ /^(384|512)$/ ? 111 : 55;
	$self->{reseed_interval} = 4294967295; # (2^32)-1
	$self->{bytes_per_request} = 2 ** 16;
	$self->{outlen} = substr($algo, -3) / 8;
	$self->{security_strength} = $self->{outlen} / 2;
	$self->{min_length} = $self->{security_strength};
	$self->{max_length} = 4294967295; # (2^32)-1

	# If we have a 64-bit Perl, make things much faster.
	my $is_64 = (4294967295 + 2) != 1;
	if ($is_64) {
		$self->{s_add} = \&_add_64;
	}
	else {
		require Math::BigInt;
		eval { Math::BigInt->import(try => 'GMP') };

		$self->{s_mask} =
			(Math::BigInt->bone << ($self->{seedlen} * 8)) - 1;
		$self->{s_add} = \&_add_32;
	}

	$self->initialize(%params);

	return $self;
}

sub _add {
	my ($self, @args) = @_;
	my $func = $self->{s_add};
	return $self->$func(@args);
}

sub _derive {
	my ($self, $hashdata, $len) = @_;

	my $count = ($len + ($self->{outlen} - 1)) / $self->{outlen};
	my $data = '';
	my $func = $self->{s_func};
	for (1..$count) {
		$data .= $func->(pack('CN', $_, $len * 8) . $hashdata);
	}
	return substr($data, 0, $len);
}

sub _seed {
	my ($self, $seed) = @_;

	my $v = $self->_derive($seed, $self->{seedlen});
	my $c = $self->_derive("\x00$v", $self->{seedlen});
	$self->{state} = {c => $c, v => $v};
	$self->{reseed_counter} = 1;
	return 1;
}

sub _reseed {
	my ($self, $seed) = @_;

	return $self->_seed("\x01$self->{state}{v}$seed");
}

sub _add_32 {
	my ($self, @args) = @_;
	my @items = map { Math::BigInt->new("0x" . unpack("H*", $_)) } @args;
	my $final = Math::BigInt->bzero;
	foreach my $val (@items) {
		$final += $val;
	}
	$final &= $self->{s_mask};
	my $data = substr($final->as_hex, 2);
	$data = "0$data" if length($data) & 1;
	$data = pack("H*", $data);
	return ("\x00" x ($self->{seedlen} - length($data))) . $data;
}

sub _add_64 {
	my ($self, $x, @args) = @_;

	use integer;

	my $nbytes = $self->{seedlen} + 1;
	my $nu32s = $nbytes / 4;
	# Optimize based on the fact that the first argument is always full-length.
	my @result = unpack('V*', reverse "\x00$x");
	my @vals = map {
		[unpack('V*', reverse(("\x00" x ($nbytes - length($_))) .  $_))]
	} @args;

	foreach my $i (0..($nu32s-1)) {
		my $total = $result[$i];
		foreach my $val (@vals) {
			$total += $val->[$i];
		}
		$result[$i+1] += $total >> 32;
		$result[$i] = $total;
	}
	return substr(reverse(pack("V*", @result[0..($nu32s-1)])), 1);
}

sub _hashgen {
	my ($self, $v, $len) = @_;

	my $func = $self->{s_func};
	my $count = int(($len + ($self->{outlen} - 1)) / $self->{outlen});
	my $data = '';
	for (1..$count) {
		$data .= $func->($v);
		$v = $self->_add($v, "\x01");
	}
	return substr($data, 0, $len);
}

=head2 $drbg->generate($bytes, $additional_data)

Generate and return $bytes bytes.  $bytes cannot exceed 2^16.

If $additional_data is specified, add this additional data to the DRBG.

=cut

sub _generate {
	my ($self, $len, $seed) = @_;

	$self->_check_reseed($len);

	my ($func, $add) = @{$self}{qw/s_func s_add/};
	my ($c, $v) = @{$self->{state}}{qw/c v/};
	if (defined $seed) {
		my $w = $func->("\x02$v$seed");
		$v = $self->$add($v, $w);
	}
	my $data = $self->_hashgen($v, $len);
	my $h = $func->("\x03$v");
	$v = $self->$add($v, $h, $c, pack("N*", $self->{reseed_counter}));
	$self->{reseed_counter}++;
	$self->{state}{v} =  $v;
	return substr($data, 0, $len);
}

=head1 AUTHOR

brian m. carlson, C<< <sandals at crustytoothpaste.net> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-crypt-drbg at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Crypt-DRBG>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc Crypt::DRBG::Hash


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-DRBG>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Crypt-DRBG>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Crypt-DRBG>

=item * Search CPAN

L<http://search.cpan.org/dist/Crypt-DRBG/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2015 brian m. carlson.

This program is distributed under the MIT (X11) License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.


=cut

1; # End of Crypt::DRBG::Hash


lib/Crypt/DRBG.pm
package Crypt::DRBG;

use 5.006;
use strict;
use warnings;

use IO::File ();

=head1 NAME

Crypt::DRBG - Base class for fast, cryptographically-secure PRNGs

=head1 SYNOPSIS

	use Crypt::DRBG::HMAC;

	my $drbg = Crypt::DRBG::HMAC->new(auto => 1);
	my $data = $drbg->generate(42);
	... # do something with your 42 bytes here

	my $drbg2 = Crypt::DRBG::HMAC->new(seed => "my very secret seed");
	my @randdigits = $drbg->randitems(20, [0..9]);
	... # do something with your 20 random digits here

=head1 DESCRIPTION

Crypt::DRBG is a collection of fast, cryptographically-secure PRNGs
(pseudo-random number generators).  It can be useful for a variety of
situations:

=over 4

=item *

Cryptographically secure random numbers are needed in production, but for
testing reproducibility is needed

=item *

A large number of random values are needed, but using /dev/urandom (or the
equivalent) frequently or persistently is unsuitable

=item *

Selection of random values in a range (e.g. digits, letters, identifiers) is
required and biasing the results is unacceptable

=back

Crypt::DRBG::HMAC is the recommended class to use, as it's currently the
fastest.  All algorithms are assumed to provide equivalent security.

=head1 SUBROUTINES/METHODS

=head2 initialize(%params)

%params can contain the following:

=over 4

=item auto

If true, use a safe, cryptographically-secure set of defaults.
Equivalent to specifying autoseed, autononce, autopersonalize, and
fork_safe.

=item autoseed

If true, derive a seed from Crypt::URandom, if available,
or from /dev/urandom, /dev/arandom, or /dev/random, in that order.
Windows support requires Crypt::URandom to function properly.

=item seed

If a string, use this value as the seed.  If a coderef, call this coderef with a
single argument (the number of bytes) to obtain an entropy input.  Note that if
a string is used, an exception will be thrown if a reseed is required.

=item autononce

If true, derive a nonce automatically.

=item nonce

If a string, use this value as the nonce.  If a coderef, call this coderef with
a single argument (the number of bytes) to obtain a nonce.

=item autopersonalize

If true, derive a personalization string automatically.

=item personalize

If a string, use this value as the personalization string.  If a coderef, call
this coderef to obtain a personalization string.

=item fork_safe

If true, reseed on fork.  If false, the parent and child processes will produce
the same sequence of bytes (not recommended).

=item cache

If enabled, keep a cache of this many bytes and use it to satisfy requests
before generating more.

=back

=cut

# Not a method call.
sub _rand_bytes {
	my ($len) = @_;

	my $data = eval {
		require Crypt::URandom;
		Crypt::URandom::urandom($len);
	};

	return $data if defined $data;

	$data = '';

	my @sources = qw{/dev/urandom /dev/arandom /dev/random};
	foreach my $source (@sources) {
		my $fh = IO::File->new($source, 'r') or next;
		while ($fh->read(my $buf, $len - length($data))) {
			$data .= $buf;
		}
		die "Insufficient random data" if length($data) != $len;
		return $data;
	}
	die "No random source for autoseed";
}

sub _get_seed {
	my ($self, $name, $len, $params, $optional) = @_;
	my $autoname = "auto$name";

	my $seed;
	if (defined $params->{$name} && !ref $params->{$name}) {
		$seed = $params->{$name};
	}
	else {
		my $seedfunc;
		$seedfunc = $params->{$name} if ref $params->{$name} eq 'CODE';
		$seedfunc = \&_rand_bytes if $params->{$autoname} || $params->{auto};
		unless ($seedfunc) {
			die "No seed source" unless $optional;
			return '';
		}
		$self->{"${name}func"} = $seedfunc;
		$seed = $seedfunc->($len);
	}

	return $seed;
}

sub _get_personalization {
	my ($self, $params) = @_;
	my $name = 'personalize';
	my $autoname = "auto$name";

	my $seed;
	if (defined $params->{$name} && !ref $params->{$name}) {
		$seed = $params->{$name};
	}
	else {
		my $seedfunc;
		$seedfunc = $params->{$name} if ref $params->{$name} eq 'CODE';
		if ($params->{$autoname} || $params->{auto}) {
			die "Invalid version"
				if defined $params->{version} && $params->{version};
			my $version = 0;
			$seedfunc = sub {
				my @nums = ($$, $<, $>, time);
				my @strings = ($0, $(, $));

				return join('',
					"$version\0",
					pack("N" x @nums, @nums),
					pack("Z" x @strings, @strings),
				);
			};
		}
		# Personalization strings are recommended, but optional.
		return '' unless $seedfunc;
		$seed = $seedfunc->();
	}

	return $seed;
}

sub _check_reseed {
	my ($self) = @_;

	my $reseed = 0;
	my $pid = $self->{pid};
	$reseed = 1 if defined $pid && $pid != $$;
	$reseed = 1 if $self->{reseed_counter} >= $self->{reseed_interval};

	if ($reseed) {
		die "No seed source" if !$self->{seedfunc};
		$self->_reseed($self->{seedfunc}->($self->{seedlen}));
		$self->{pid} = $$ if $self->{fork_safe};
		$self->{cache} = '' if defined $self->{cache};
	}

	return 1;
}

sub initialize {
	my ($self, %params) = @_;

	my $seed = $self->_get_seed('seed', $self->{seedlen}, \%params);
	my $nonce = $self->_get_seed('nonce', int(($self->{seedlen} / 2) + 1),
		\%params, 1);
	my $personal = $self->_get_personalization(\%params);

	$self->_seed("$seed$nonce$personal");

	if ($params{cache}) {
		$self->{cache} = '';
		$self->{cache_size} = $params{cache};
	}

	if ($params{fork_safe} || (!exists $params{fork_safe} && $params{auto})) {
		$self->{fork_safe} = 1;
		$self->{pid} = $$;
	}

	return 1;
}

=head2 $drbg->generate($bytes, $additional_data)

Generate and return $bytes bytes.  There is a limit per algorithm on the number of bytes that can be requested at once, which is at least 2^10.

If $additional_data is specified, add this additional data to the DRBG.

If the cache flag was specified on instantiation, bytes will be satisfied from
the cache first, unless $additional_data was specified.

=cut

sub generate {
	my ($self, $len, $seed) = @_;

	return $self->_generate($len, $seed)
		if !defined $self->{cache} || defined $seed;

	$self->_check_reseed;

	my $data = '';
	my $left = $len;
	my $cache = \$self->{cache};
	$$cache = $self->_generate($self->{cache_size}) if !length($$cache);
	while ($left > 0) {
		my $chunk_size = $left > length($$cache) ? length($$cache) : $left;
		$data .= substr($$cache, 0, $chunk_size, '');
		$left = $len - length($data);
		$$cache = $self->_generate($self->{cache_size}) if !length($$cache);
	}

	return $data;
}

=head2 $drbg->rand([$n], [$num])

Like Perl's rand, but cryptographically secure.  Uses 32-bit values.

Accepts an additional argument, $num, which is the number of values to return.
Defaults to 1 (obviously).

Note that just as with Perl's rand, there may be a slight bias with this
function.  Use randitems if that matters to you.

Returns an array if $num is specified and a single item if it is not.

=cut

sub rand {
	my ($self, $n, $num) = @_;

	my $single = !defined $num;

	$n = 1 unless defined $n;
	$num = 1 unless defined $num;

	my $bytes = $self->generate($num * 4);
	my @data = map { $_ / 2.0 / (2 ** 31) * $n } unpack("N[$num]", $bytes);
	return $single ? $data[0] : @data;
}

=head2 $drbg->randitems($n, $items)

Select randomly and uniformly from the arrayref $items $n times.

=cut

sub randitems {
	my ($self, $n, $items) = @_;

	my $len = scalar @$items;
	my @results;
	my $values = [
		{bytes => 1, pack => 'C', max => 256},
		{bytes => 2, pack => 'n', max => 65536},
		{bytes => 4, pack => 'N', max => 2 ** 31},
	];
	my $params = $values->[$len <= 256 ? 0 : $len <= 65536 ? 1 : 2];

	# Getting this computation right is important so as not to bias the
	# data.  $len & $len - 1 is true iff $len is not a power of two.
	my $max = $params->{max};
	my $mask = $max - 1;
	if ($len & ($len - 1)) {
		$max = $max - ($max % $len);
	}
	else {
		$mask = $len - 1;
	}

	my $pack = "$params->{pack}\[$n\]";
	while (@results < $n) {
		my $bytes = $self->generate($params->{bytes} * $n);

		my @data = map { $_ & $mask } grep { $_ < $max } unpack($pack, $bytes);
		push @results, map { $items->[$_ % $len] } @data;
	}

	return splice(@results, 0, $n);
}

=head2 $drbg->randbytes($n, $items)

Select randomly and uniformly from the characters in arrayref $items $n times.
Returns a byte string.

This function works just like randitems, but is more efficient if generating a
sequence of bytes as a string instead of an array.

=cut

sub randbytes {
	my ($self, $n, $items) = @_;

	my $len = scalar @$items;
	my $results = '';

	# Getting this computation right is important so as not to bias the
	# data.  $len & $len - 1 is true iff $len is not a power of two.
	my $max = 256;
	my $filter = sub { return $_[0]; };
	if ($len & ($len - 1)) {
		$max = $max - ($max % $len);
		my $esc = sprintf '\x%02x', $max + 1;
		$filter = sub {
			my $s = shift;
			eval "\$s =~ tr/$esc-\\xff//d";  ## no critic(ProhibitStringyEval)
			return $s;
		};
	}

	while (length $results < $n) {
		my $bytes = $filter->($self->generate($n));
		$results .= join '', map { $items->[$_ % $len] } unpack('C*', $bytes);
	}

	return substr($results, 0, $n);
}

=head1 AUTHOR

brian m. carlson, C<< <sandals at crustytoothpaste.net> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-crypt-drbg at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Crypt-DRBG>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc Crypt::DRBG


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-DRBG>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Crypt-DRBG>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Crypt-DRBG>

=item * Search CPAN

L<http://search.cpan.org/dist/Crypt-DRBG/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2015 brian m. carlson.

This program is distributed under the MIT (X11) License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.


=cut

1; # End of Crypt::DRBG
